<a href="https://colab.research.google.com/github/hamidrezanorouzi/numericalMethods/blob/main/Lectures/Lecture05_linear_systems_Part2.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

# **System of Linear Equations - Part 2**

&nbsp;

Lecturer: **Hamidreza Norouzi**

&nbsp;

### **Notes**❕
This content forms a part of the instructional presentations for the **`numerical methods in chemical engineering`** course designed for undergraduate chemical engineering students at Amirkabir University of Technology.

Feel free to utilize the information and source codes provided in this material, ensuring appropriate acknowledgment of the original document.

The visual elements featured in this document are either original or have been obtained from the following sources, unless specified otherwise:

* Steven C. Chapra, Applied Numerical Methods with Matlab for Engineers and Scientists, 3rd edition, McGraw-Hill (2012).
* Amos Gilat and Vish Subramanian, Numerical Methods for Engineers and Scientists, 3rd edition, Wiley (2014).


<div align="center">
🟧 🟧 🟧
</dive>

---

# 🔵 1) LU factorization

* Consider the following system:
$$
[A]\{x\}=\{b\} \tag{1-1}
$$
* It happens in engineering applications that the matrix A is not changing and only B changes for solving the problems.
* In this condition, Gauss elimination becomes inefficient.
* We can re-arrange Eq. (1-1) into:
$$
[A]\{x\}-\{b\}=0 \tag{1-2}
$$
* Consider that Eq. (1-2) can be converted into upper triangular system, for example for a $3\times 3$ system:
$$
\begin{bmatrix}
u_{11} & u_{12} & u_{13} \\
0 & u_{22} & u_{23} \\
0 & 0 & u_{33}
\end{bmatrix}
\begin{bmatrix}
x_1 \\ x_2 \\x_3
\end{bmatrix} =
\begin{bmatrix}
d_1 \\ d_2 \\ d_3
\end{bmatrix}
\tag{1-3}
$$


* Eq. (1-3) can be shown in compact version as:
$$
[U]\{x\}-\{d\} = 0 \tag{1-4}
$$
* Now suppose that there is lower triangular matrix like this:
$$
L=
\begin{bmatrix}
1 & 0 & 0\\
l_{21} & 1 & 0 \\
l_{31} & l_{32} & 1
\end{bmatrix}
\tag{1-5}
$$
 that has the property that when Eq. (1-4) is premultiplied by it, Eq. (1-2) is the result.
$$
[L]\{[U]\{x\}-\{d\}\} = [A]\{x\}-\{b\}  \tag{1-6}
$$


* So we can say:
 $$
 [L][U] = [A]
 \tag{1-7}
 $$

 &nbsp;

 $$
 [L]\{d\}= \{b\} \\
 d_i = b_i - \sum_{j=1}^{i-1} l_{ij}d_j
 \tag{1-8}
 $$

* We can use a two-step strategy for solving such a system:
 * **Step1:** LU factorization in which $[A]$ is decomposed into $[U]$ and $[L]$.
 * **Step2:** Obtain $\{d\}$ from Eq. (1-8) and then obtain $\{x\}$ from Eq. (1-4).

<div align="center">
<img src="https://drive.google.com/uc?id=1_EC4mDmwyxRjk8a_J6ygMLdguelvKFer" width="500">
</div>

## 1-1) How to factorize?
* Factorization is performed using the row operations that we used in Gauss elimination.
* We show the procedure for a $3\times 3$ matrix, and the procedure can be extended to larger systems too.
* In the process of converting the coefficient matrix into an upper-triangular matrix, we use row operations:
$$
\begin{bmatrix}
a_{11}& a_{12}& a_{13} \\
a_{21}& a_{22}& a_{23} \\
a_{31}& a_{32}& a_{33} \\
\end{bmatrix}
$$
* Starting with pivot element $a{11}$:
  * Row1 is multiplied with $f_{21}$ and subtracted from Row2 to eliminate $a_{21}$:
  $$
  f_{21}=\frac{a_{21}}{a_{11}}
  $$
  * Row1 is multiplied with $f_{31}$ and subtracted from Row3 to eliminate $a_{31}$:
  $$
  f_{31}=\frac{a_{31}}{a_{11}}
  $$
  * The result is:
  $$
 \begin{bmatrix}
 a_{11}& a_{12}& a_{13} \\
 0& a^{'}_{22}& a^{'}_{23} \\
 0& a^{'}_{32}& a^{'}_{33} \\
 \end{bmatrix}
 $$


* Starting with pivot element $a^{'}_{22}$:
  * Row2 is multiplied with $f_{32}$ and subtracted from Row3 to eliminate $a^{'}_{32}$:
  $$
  f_{32} = \frac{a^{'}_{32}}{a^{'}_{22}}
  $$
  * The result is:
  $$
  [U]=
  \begin{bmatrix}
  a_{11}& a_{12}& a_{13} \\
  0& a^{'}_{22}& a^{'}_{23} \\
  0& 0& a^{"}_{33} \\
  \end{bmatrix}
  $$
* Now if we store factors in a separate matrix $[L]$, we can show:
$$
[A] = [L][U]
$$

&nbsp;

$$
[L] =
\begin{bmatrix}
1 & 0& 0 \\
f_{21} & 1 & 0 \\
f{31} & f_{32} & 0
\end{bmatrix}
$$

### ❓ **Example1:**

Consider the following system of equations, use LU factorization to solve it.
$$
\begin{cases}
3x_1 -0.1x_2-0.2x_3 = 7.85 \\
0.1x_1 + 7x_2-0.3x_3 = -19.3 \\
0.3x_1 - 0.2x_2+10x_3 = 71.4
\end{cases}
$$

💡 *Solution*

The coefficient matrix for this system is:
$$
A =
\begin{bmatrix}
3 & -0.1 & -0.2 \\
0.1 & 7 & -0.3 \\
0.3 & -0.2 & 10
\end{bmatrix}
$$
**Forward elimination**:

Taking 3 as the pivot element and zeroing $a_{21}$ and $a_{31}$:
$$
f_{21} = \frac{0.1}{3} = 0.0333333 \\
f_{31} = \frac{0.3}{3} = 0.100000
$$
\begin{bmatrix}
3 & -0.1 & -0.2 \\
0 & 7.00333 & -0.293333 \\
0 & -0.19 & 1.02
\end{bmatrix}


Taking 7.00333 as the pivot element and zeroing $a^{'}_{32}$:
$$
f_{32} = \frac{-0.19}{7.00333} = -0.0271300
$$
$$
[U] =
\begin{bmatrix}
3 & -0.1 & -0.2 \\
0 & 7.00333 & -0.293333 \\
0 & 0 & 10.0120
\end{bmatrix}
$$
and [L] matrix becomes:
$$
[L]=
\begin{bmatrix}
1 & 0 & 0 \\
0.0333333 & 1 & 0 \\
0.1 & -0.0271300 & 1
\end{bmatrix}
$$
We can test the factorized matrices:
$$
[L][U] =
\begin{bmatrix}
3.00000 & -0.100000 & -0.200000 \\
0.0999999 & 7.00000 & -0.300000 \\
3.00000 & -0.200000 & 9.99996
\end{bmatrix}
$$

**Forward-substitution**
$$
[L]\{d\}=\{b\}
$$
we can re-write the above equation as:

$$
\begin{aligned}
d_i &= b_i - \sum_{j=1}^{i-1} l_{ij}d_j \\
⇒ d_1 &= 7.85 \\
d_2 &= -19.3 - 0.0333333 d_1 = -19.3 - 0.0333333(7.85) = -19.5617 \\
d_3 &= 71.4 - 0.1d_1 + 0.02713d_2 = 71.4 -0.1(7.85) + 0.02713(-19.5617) = 70.0843
\\
⇒
\{d\} &=
\begin{bmatrix}
7.85 \\ -19.5617 \\ 70.0843
\end{bmatrix}
\end{aligned}
$$



**Back-substitution**

We can find the final solution using back-substitution:
$$
[U]\{x\}=\{d\}
$$

$$
\begin{bmatrix}
3 & -0.1 & -0.2 \\
0 & 7.00333 & -0.293333 \\
0 & 0 & 10.0120
\end{bmatrix}
\begin{bmatrix}
x_1 \\ x_2 \\ x_3
\end{bmatrix}
=
\begin{bmatrix}
7.85 \\ -19.5617 \\ 70.0843
\end{bmatrix}
⇒ \{x\} =
 \begin{bmatrix}
3.00000 \\ -2.50000 \\ 7.00003
\end{bmatrix}
$$


## 1-2) Code for LU factorization

In [1]:
import numpy as np

def lu(matrix):

  n = len(matrix)
  L = np.identity(n)
  U = matrix.copy().astype('double')

  for k in range(n - 1):

    # factorization
    for i in range(k + 1, n):
      factor = U[i, k] / U[k, k]
      L[i, k] = factor
      U[i, k:] -= factor * U[k, k:]

  return L, U

def solveUsingLU(L, U, b):
  n = len(L)
  d=np.empty((n,))
  d[0] = b[0]

  # forward sub
  for i in range(1,n):
    d[i] = b[i] - np.dot(d[:i],L[i,:i]);

  # back sub
  x = np.empty((n,))
  x[n-1] = d[n-1]/U[n-1,n-1]
  for i in range(n-2,-1,-1):
    x[i] = (d[i] - np.dot(U[i,i+1:],x[i+1:]))/U[i,i]

  return x

In [2]:
A = np.array([[3, -0.1, -0.2],[0.1, 7, -0.3],[0.3, -0.2, 10]])
b = np.array([7.85, -19.3, 71.4])
L, U = lu(A)
x = solveUsingLU(L,U,b)
print(L)
print(U)

print("The solution is \n", x)

[[ 1.          0.          0.        ]
 [ 0.03333333  1.          0.        ]
 [ 0.1        -0.02712994  1.        ]]
[[ 3.         -0.1        -0.2       ]
 [ 0.          7.00333333 -0.29333333]
 [ 0.          0.         10.01204188]]
The solution is 
 [ 3.  -2.5  7. ]


## 1-3) LU factorization with pivoting
* Pivoting in inevitable with Gauss elimination to obtain a better solution.
* We use **permutation matrix** to keep track of row changes during forward elimination phase (will be explained later).
* With permutation matrix $[P]$, the set of equations become:
  * After factorization we have  
  $$
  [P][A] = [L][U] \tag{1-9}
  $$
  * Forward substitution is performed to obtain vector $\{d\}$:
  $$
  [L]\{d\} = [P]\{b\} \tag{1-10}
  $$
  * Back-substitution is performed to obtain the final solution:
  $$
  [U]\{x\} = \{d\} \tag{1-11}
  $$

## 1-4) What is permutation matrix and how to calculate it?
* Permutation matrix is an identity matrix whose rows/columns are interchanged:
$$
\begin{bmatrix}
1&0&0 \\
0&1&0 \\
0&0&1
\end{bmatrix}
⇒ \text{interchange of rows 1 and 3}  ⇒
\begin{bmatrix}
0&0&1 \\
0&1&0 \\
1&0&0 \\
\end{bmatrix}
$$
* If it $[A]$ is pre-multiplied by $[P]$, the corresponding rows/columns of $[A]$ are interchanged.

&nbsp;

$$
[P][A] = \begin{bmatrix}
0&0&1 \\
0&1&0 \\
1&0&0 \\
\end{bmatrix}
\begin{bmatrix}
11&12&13 \\
21&22&23 \\
31&32&33 \\
\end{bmatrix}
=
\begin{bmatrix}
31&32&33 \\
21&22&23 \\
11&12&13 \\
\end{bmatrix}
$$

### ❓ **Example 2:**
Consider the following system of equations, use LU factorization to solve it.
$$
\begin{cases}
0.3x_1 - 0.2x_2+10x_3 = 71.4 \\
0.1x_1 + 7x_2-0.3x_3 = -19.3 \\
3x_1 -0.1x_2-0.2x_3 = 7.85 \\
\end{cases}
$$

💡 *Solution*

The coefficient matrix for this system is:
$$
A =
\begin{bmatrix}
0.3 & -0.2 & 10 \\
0.1 & 7 & -0.3 \\
3 & -0.1 & -0.2 \\
\end{bmatrix}
$$
**Forward elimination**:

Starting with $a_{11}$, we find out that it is not the largest element in the first column, so we interchange the rows 3 and 1 and we obtain:

$$
[A^{'}] =
\begin{bmatrix}
3 & -0.1 & -0.2 \\
0.1 & 7 & -0.3 \\
0.3 & -0.2 & 10 \\
\end{bmatrix}
$$

and the permutation matrix becomes:
$$
[P] =
\begin{bmatrix}
0&0&1 \\
0&1&0 \\
1&0&0 \\
\end{bmatrix}
$$

Matrix $[A^{'}]$ is similar to what we had originally in the previous example, so, after factorization, we will have:

$$
\begin{aligned}
L &=
\begin{bmatrix}
1 & 0 & 0 \\
0.0333333 & 1 & 0 \\
0.1 & -0.0271300 & 1
\end{bmatrix}
\\
U &=
\begin{bmatrix}
3 & -0.1 & -0.2 \\
0 & 7.00333 & -0.293333 \\
0 & 0 & 10.0120
\end{bmatrix}
\end{aligned}
$$



**Forward-substitution**
In this step with permutation matrix, we use Eq. (1-10):
$$
[L]\{d\}=[P]\{b\}
$$

&nbsp;

$$
[Pb] =
\begin{bmatrix}
0&0&1 \\
0&1&0 \\
1&0&0 \\
\end{bmatrix}
\begin{bmatrix}
71.4 \\ -19.3 \\ 7.85
\end{bmatrix}
=
\begin{bmatrix}
7.85 \\ -19.3 \\ 71.4
\end{bmatrix}
$$

&nbsp;

now we can obtain vector $\{d\}$:

$$
\begin{aligned}
d_i &= Pb_i - \sum_{j=1}^{i-1} l_{ij}d_j \\
⇒ d_1 &= 7.85 \\
d_2 &= -19.3 - 0.0333333 d_1 = -19.3 - 0.0333333(7.85) = -19.5617 \\
d_3 &= 71.4 - 0.1d_1 + 0.02713d_2 = 71.4 -0.1(7.85) + 0.02713(-19.5617) = 70.0843
\\
⇒
\{d\} &=
\begin{bmatrix}
7.85 \\ -19.5617 \\ 70.0843
\end{bmatrix}
\end{aligned}
$$



**Back-substitution**

We can find the final solution using back-substitution:

&nbsp;

$$
[U]\{x\}=\{d\}
$$

$$
\begin{bmatrix}
3 & -0.1 & -0.2 \\
0 & 7.00333 & -0.293333 \\
0 & 0 & 10.0120
\end{bmatrix}
\begin{bmatrix}
x_1 \\ x_2 \\ x_3
\end{bmatrix}
=
\begin{bmatrix}
7.85 \\ -19.5617 \\ 70.0843
\end{bmatrix}
⇒ \{x\} =
 \begin{bmatrix}
3.00000 \\ -2.50000 \\ 7.00003
\end{bmatrix}
$$


## 1-5) Code for LU factorization with pivoting

In [3]:
import numpy as np

def luPivoting(matrix):
  n = len(matrix)

  L = np.identity(n)
  U = matrix.copy().astype('double')
  P = np.identity(n)

  for k in range(n - 1):

    # Partial Pivoting
    pivot_row = np.argmax(abs(U[k:, k])) + k
    if pivot_row != k:
      U[[k, pivot_row]] = U[[pivot_row, k]]
      P[[k, pivot_row]] = P[[pivot_row, k]]
      if k > 0:
        L[[k, pivot_row], :k] = L[[pivot_row, k], :k]

    # factorization
    for i in range(k + 1, n):
      factor = U[i, k] / U[k, k]
      L[i, k] = factor
      U[i, k:] -= factor * U[k, k:]

  return P, L, U


def solveUsingPLU(P, L, U, b):
  n = len(L)
  d=np.empty((n,))

  Pb = np.dot(P,b)

  d[0] = Pb[0]

  # forward sub
  for i in range(1,n):
    d[i] = Pb[i] - np.dot(d[:i],L[i,:i]);

  # back sub
  x = np.empty((n,))
  x[n-1] = d[n-1]/U[n-1,n-1]
  for i in range(n-2,-1,-1):
    x[i] = (d[i] - np.dot(U[i,i+1:],x[i+1:]))/U[i,i]

  return x

In [4]:
A = np.array([[0.3, -0.2, 10], [0.1, 7, -0.3],[3, -0.1, -0.2]])
b = np.array([71.4, -19.3, 7.85])
P, L, U = luPivoting(A)
x = solveUsingPLU(P,L,U,b)

print("The solution is \n", x)

The solution is 
 [ 3.  -2.5  7. ]


<div align="center">
🟦 🟦 🟦
</div>


---



# 🔴 2) Gauss-Seidel method

* Certain problems in engineering produce a set of linear equations which are dominantly diagonal systems.

* **Dominantly diagonal matrix**: in each row, the absolute value of the diagonal element is greater than the sum of the absolute values of off-diagonal elements.
$$
|a_{ii}|>\sum_{j=1, j\neq i}^n|a_{ij}|
\tag{2-1}
$$

* We can use iterative method for solving such a system.


* For example, consider the following system of equations:
$$
\begin{cases}
x_1+8x_2-2x_3 = 9  \\
-3x_1-x_2-7x_3 = -33 \\
-10x_1+2x_2+3x_3=6 \\
\end{cases}
$$
* First, we must rearrange rows / equations to obtain a system that most/all of the rows are diagonal.
\begin{cases}
-10x_1+2x_2+3x_3=6 \\
x_1+8x_2-2x_3 = 9  \\
-3x_1-x_2-7x_3 = -33 \\
\end{cases}
* Each equation can be solved for its diagonal element:
$$
\begin{aligned}
x_1 &= \frac{6-2x_2-3x_3}{-10} &(Eq1)\\
x_2 &= \frac{ 9-x_1+2x_3}{8} &(Eq2)\\
x_3 &= \frac{-33+3x_1+x_2}{-7} &(Eq3)\\
\end{aligned}
$$

* To solve these equations, an initial guess is required for $x^0_2$ and $x^0_3$.
* Using values of $x^0_2$ and $x^0_3$ a new estimate of $x^1_1$ is obtained from (Eq1).
* Using values of $x^0_3$ and $x^1_1$, a new estimate of $x^1_2$ is obtained from (Eq2).
* Using values of $x^1_1$ and $x^1_2$, a new estimate of $x^1_3$ is obtained from (Eq3).

* The estimated value of $\{x\}$ is then used as the guess in the next iteration.
* Iteration continues until a convergence occurs:
$$
|\frac{x^j_i-x^{j-1}_i}{x^j_i}|\times 100\% < tol \ \ \ \text{for} \ \ \ i=1,2,...,n
\tag{2-2}
$$

### ❓ **Example 3:**
Use Gauss-Seidel method, obtain the solution for the following system with initial guess, $x_1=0$, $x_2=0$, $x_3=0$ and $tol = 0.1%$.

$$
\begin{cases}
0.1x_1 + 7x_2-0.3x_3 = -19.3 \\
0.3x_1 - 0.2x_2+10x_3 = 71.4 \\
3x_1 -0.1x_2-0.2x_3 = 7.85
\end{cases}
$$

💡 *Solution*

We first arrange equations based on the element with largest coefficient.
$$
\begin{aligned}
x_1 &= \frac{7.85+0.1x_2+0.2x_3}{3} \\
x_2 &= \frac{-19.3-0.1x_1+0.3x_3}{7} \\
x_3 &= \frac{71.4-0.3x_1+0.2x_2}{10}
\end{aligned}
$$
**iteration1:**
\begin{cases}
x^0_2 = 0, x^0_3=0 &⇒ x^1_1 &= \frac{7.85+0.1(0)+0.2(0)}{3} = 2.6167 \\
x^1_1 = 2.6167, x^0_3=0 &⇒ x^1_2 &= \frac{-19.3-0.1(2.6167)+0.3(0)}{7} = -2.7945 \\
x^1_1 = 2.6167, x^1_2=-2.7945 &⇒ x^1_3 &= \frac{71.4-0.3(2.6167)+0.2(-2.7945)}{10} = 7.0056
\end{cases}

**iteration2:**
\begin{cases}
x^1_1 = 2.6167 \\
x^1_2 = -2.7945, x^1_3= 7.0056 &⇒ x^2_1 &= \frac{7.85+0.1(-2.7945)+0.2(7.0056)}{3} = 2.9906 \\
x^2_1 = 2.9906, x^1_3=7.00560 &⇒ x^2_2 &= \frac{-19.3-0.1(2.9906)+0.3(7.0056)}{7} = -2.4996 \\
x^2_1 = 2.9906, x^2_2=-2.4996 &⇒ x^2_3 &= \frac{71.4-0.3(2.9906)+0.2(-2.4996)}{10} = 7.0003
\end{cases}



**iteration3:**
\begin{cases}
x^2_1 = 2.9906 \\
x^2_2 = -2.4996, x^2_3= 7.0003 &⇒ x^3_1 &= \frac{7.85+0.1(-2.4996)+0.2(7.0003)}{3} = 3.0000 \\
x^3_1 = 3.0000, x^2_3=7.0003 &⇒ x^3_2 &= \frac{-19.3-0.1(3.0000)+0.3(7.0003)}{7} = -2.5000 \\
x^3_1 = 3.0000, x^3_2=-2.5000 &⇒ x^3_3 &= \frac{71.4-0.3(3.0000)+0.2(-2.5000)}{10} = 7.0000
\end{cases}

\begin{cases}
ϵ_{a1} = |\frac{3-2.9906}{3}|\times 100 = 0.31\% ❌\\
ϵ_{a2} = |\frac{-2.5+2.4996}{-2.5}|\times 100 = 0.016\% ✅ \\
ϵ_{a3} = |\frac{7-7.0003}{7}|\times 100 = 0.004\% ✅\\
\end{cases}

**iteration4:**

If we perform one more iteration we obtain:
$$
x^4_1 = 3.0000, x^4_2 = -2.5000, x^4_3= 7.0000
$$

## 2-1) Code for Gauss-Seidel method


In [5]:
import numpy as np

def GaussSeidel(A, b, x0, tol = 1.0e-3):
  n = len(b)
  x = x0.copy().astype('double')

  for iter in range(100):
    for i in range(n):
      sum1 = np.dot(A[i, :i], x[:i])
      sum2 = np.dot(A[i, i+1:], x[i+1:])
      x[i] = (b[i] - sum1 - sum2) / A[i, i]

    # Check for convergence
    if np.all(np.abs((x - x0)/np.maximum(x,1.0e-14)) < tol):
      return x

    x0 = x.copy()

  # no solution
  return None

In [6]:
A = np.array([[3, -0.1, -0.2],[0.1, 7, -0.3],[0.3, -0.2, 10]])
B = np.array([7.85, -19.3, 71.4])
x0 = np.array([0,0,0])

res = GaussSeidel(A,B,x0)
if res is not None:
  print("The solution is \n", res)
else:
  print("No solution!")

The solution is 
 [ 3.  -2.5  7. ]


<div align="center">
🟥 🟥 🟥
</div>

---

